In [2]:
#Importamos librerías
import pandas as pd
import re

In [2]:
#Ponemos todas las filas en un solo dataframe
for i in range(1,12):
    if i == 1:
        google_metadata = pd.read_json('metadata-sitios/'+str(i)+'.json',lines=True)
    else:
        google_metadata = pd.concat([google_metadata,pd.read_json('metadata-sitios/'+str(i)+'.json',lines=True)])
google_metadata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3025011 entries, 0 to 275000
Data columns (total 15 columns):
 #   Column            Dtype  
---  ------            -----  
 0   name              object 
 1   address           object 
 2   gmap_id           object 
 3   description       object 
 4   latitude          float64
 5   longitude         float64
 6   category          object 
 7   avg_rating        float64
 8   num_of_reviews    int64  
 9   price             object 
 10  hours             object 
 11  MISC              object 
 12  state             object 
 13  relative_results  object 
 14  url               object 
dtypes: float64(3), int64(1), object(11)
memory usage: 369.3+ MB


In [10]:
google_metadata.head()

name                                            address  \
0   Porter Pharmacy  Porter Pharmacy, 129 N Second St, Cochran, GA ...   
1      City Textile  City Textile, 3001 E Pico Blvd, Los Angeles, C...   
2      San Soo Dang  San Soo Dang, 761 S Vermont Ave, Los Angeles, ...   
3      Nova Fabrics  Nova Fabrics, 2200 E 11th St, Los Angeles, CA ...   
4  Nobel Textile Co  Nobel Textile Co, 719 E 9th St, Los Angeles, C...   

                                 gmap_id description   latitude   longitude  \
0  0x88f16e41928ff687:0x883dad4fd048e8f8        None  32.388300  -83.357100   
1  0x80c2c98c0e3c16fd:0x29ec8a728764fdf9        None  34.018891 -118.215290   
2  0x80c2c778e3b73d33:0xbdc58662a4a97d49        None  34.058092 -118.292130   
3   0x80c2c89923b27a41:0x32041559418d447        None  34.023669 -118.232930   
4  0x80c2c632f933b073:0xc31785961fe826a6        None  34.036694 -118.249421   

              category  avg_rating  num_of_reviews price  \
0           [Pharmacy]         4.9              16  None   
1   [Textile exporter]         4.5               6  None   
2  [Korean restaurant]         4.4              18  None   
3       [Fabric store]         3.3               6  None   
4       [Fabric store]         4.3               7  None   

                                               hours  \
0  [[Friday, 8AM–6PM], [Saturday, 8AM–12PM], [Sun...   
1                                               None   
2  [[Thursday, 6:30AM–6PM], [Friday, 6:30AM–6PM],...   
3  [[Thursday, 9AM–5PM], [Friday, 9AM–5PM], [Satu...   
4  [[Thursday, 9AM–5PM], [Friday, 9AM–5PM], [Satu...   

                                                MISC              state  \
0  {'Service options': ['In-store shopping', 'Sam...  Open ⋅ Closes 6PM   
1                                               None           Open now   
2  {'Service options': ['Takeout', 'Dine-in', 'De...  Open ⋅ Closes 6PM   
3  {'Service options': ['In-store shopping'], 'Pa...  Open ⋅ Closes 5PM   
4           {'Service options': ['In-store pickup']}  Open ⋅ Closes 5PM   

                                    relative_results  \
0  [0x88f16e41929435cf:0x5b2532a2885e9ef6, 0x88f1...   
1  [0x80c2c624136ea88b:0xb0315367ed448771, 0x80c2...   
2  [0x80c2c78249aba68f:0x35bf16ce61be751d, 0x80c2...   
3  [0x80c2c8811477253f:0x23a8a492df1918f7, 0x80c2...   
4  [0x80c2c62c496083d1:0xdefa11317fe870a1, 0x80c2...   

                                                 url  
0  https://www.google.com/maps/place//data=!4m2!3...  
1  https://www.google.com/maps/place//data=!4m2!3...  
2  https://www.google.com/maps/place//data=!4m2!3...  
3  https://www.google.com/maps/place//data=!4m2!3...  
4  https://www.google.com/maps/place//data=!4m2!3...

Iniciamos con 3.025.011 registros. Vamos a ir reduciendo los datos quitando todo lo que consideremos innecesario (duplicados, nulos no manejables, etc) o que no aporte información.

In [4]:
#Descartamos los resultados con mismo gmap_id que es el identificador del lugar
google_metadata.drop_duplicates(subset='gmap_id',inplace=True)
google_metadata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2998428 entries, 0 to 275000
Data columns (total 15 columns):
 #   Column            Dtype  
---  ------            -----  
 0   name              object 
 1   address           object 
 2   gmap_id           object 
 3   description       object 
 4   latitude          float64
 5   longitude         float64
 6   category          object 
 7   avg_rating        float64
 8   num_of_reviews    int64  
 9   price             object 
 10  hours             object 
 11  MISC              object 
 12  state             object 
 13  relative_results  object 
 14  url               object 
dtypes: float64(3), int64(1), object(11)
memory usage: 366.0+ MB


In [5]:
#Como solo nos vamos a dedicar a los restaurants vamos a descartar todos los registros que tengan nulos bajo el campo category
google_metadata.dropna(subset='category',inplace=True)
google_metadata.reset_index(inplace=True,drop=True)
google_metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2981245 entries, 0 to 2981244
Data columns (total 15 columns):
 #   Column            Dtype  
---  ------            -----  
 0   name              object 
 1   address           object 
 2   gmap_id           object 
 3   description       object 
 4   latitude          float64
 5   longitude         float64
 6   category          object 
 7   avg_rating        float64
 8   num_of_reviews    int64  
 9   price             object 
 10  hours             object 
 11  MISC              object 
 12  state             object 
 13  relative_results  object 
 14  url               object 
dtypes: float64(3), int64(1), object(11)
memory usage: 341.2+ MB


Redujimos hasta ahora a 2.981.245 registros. De estos vamos a filtrar los que tengan restaurant bajo el campo "category".

In [6]:
#Hallamos los índices de los restaurants
restaurant_indexes = []
for i in range(google_metadata.shape[0]):
    try:
        for j in range(len(google_metadata['category'][i])):
            if google_metadata['category'][i][j].find('estaura')!=-1:
                restaurant_indexes.append(i)
                break
    except:
        print(i)
len(restaurant_indexes)

210953

In [7]:
#Filtramos los restaurants
google = google_metadata[google_metadata.index.isin(restaurant_indexes)]
google.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 210953 entries, 2 to 2981240
Data columns (total 15 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   name              210953 non-null  object 
 1   address           209900 non-null  object 
 2   gmap_id           210953 non-null  object 
 3   description       72174 non-null   object 
 4   latitude          210953 non-null  float64
 5   longitude         210953 non-null  float64
 6   category          210953 non-null  object 
 7   avg_rating        210953 non-null  float64
 8   num_of_reviews    210953 non-null  int64  
 9   price             99738 non-null   object 
 10  hours             189013 non-null  object 
 11  MISC              209549 non-null  object 
 12  state             189848 non-null  object 
 13  relative_results  170277 non-null  object 
 14  url               210953 non-null  object 
dtypes: float64(3), int64(1), object(11)
memory usage: 25.8+ MB


De los mas de 3.000.000 de registros, solamente 210.953 son restaurants. Nos vamos a dedicar a limpiar y normalizar estos. Como primera medida vamos a normalizar los campos. Lo que buscamos es tener una tabla que contenga las siguientes columnas:

* Id_Restaurant
* Nombre
* Ciudad
* Estado
* Código postal
* Latitud
* Longitud
* Tipo (category)
* Atributos (MISC)

La información sobre la ciudad, estado y código postal se encuentra bajo el campo address del dataframe de google. Vamos a separarlo.

In [25]:
#Quitamos los valores que no tengan el valor de address y los copiamos en otro dataframe
restaurants_google = google.dropna(subset='address')
restaurants_google.reset_index(inplace=True, drop=True)
restaurants_google.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209900 entries, 0 to 209899
Data columns (total 15 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   name              209900 non-null  object 
 1   address           209900 non-null  object 
 2   gmap_id           209900 non-null  object 
 3   description       72125 non-null   object 
 4   latitude          209900 non-null  float64
 5   longitude         209900 non-null  float64
 6   category          209900 non-null  object 
 7   avg_rating        209900 non-null  float64
 8   num_of_reviews    209900 non-null  int64  
 9   price             99601 non-null   object 
 10  hours             188226 non-null  object 
 11  MISC              208666 non-null  object 
 12  state             189058 non-null  object 
 13  relative_results  169451 non-null  object 
 14  url               209900 non-null  object 
dtypes: float64(3), int64(1), object(11)
memory usage: 24.0+ MB


In [12]:
#Miramos el dataframe resultante
restaurants_google.head()

name  \
0                           San Soo Dang   
1                           Vons Chicken   
2  Sweet Rewards Gluten Free Bakery, LLC   
3                        Vivi Bubble Tea   
4                              Hale Pops   

                                             address  \
0  San Soo Dang, 761 S Vermont Ave, Los Angeles, ...   
1  Vons Chicken, 12740 La Mirada Blvd, La Mirada,...   
2  Sweet Rewards Gluten Free Bakery, LLC, 85 NE D...   
3  Vivi Bubble Tea, 701 S 5th St, Philadelphia, P...   
4   Hale Pops, 55-370 Kamehameha Hwy, Laie, HI 96762   

                                 gmap_id  \
0  0x80c2c778e3b73d33:0xbdc58662a4a97d49   
1  0x80dd2b4c8555edb7:0xfc33d65c4bdbef42   
2   0x87ec235c54d25b31:0x3b75fb5facc602f   
3  0x89c6c89efcaed69d:0xded973f6033e7dba   
4   0x7c00456eecad3111:0x8217f9600c51f33   

                                         description   latitude   longitude  \
0                                               None  34.058092 -118.292130   
1                                               None  33.916402 -118.010855   
2                                               None  41.616079  -93.865487   
3  Bubble tea shop offering hot & cold versions, ...  39.940293  -75.150923   
4                                               None  21.637796 -157.920714   

                           category  avg_rating  num_of_reviews price  \
0               [Korean restaurant]         4.4              18  None   
1                      [Restaurant]         4.5              18  None   
2  [Bakery, Health food restaurant]         4.7              21  None   
3                      [Restaurant]         4.0               8  None   
4                      [Restaurant]         4.4              18  None   

                                               hours  \
0  [[Thursday, 6:30AM–6PM], [Friday, 6:30AM–6PM],...   
1  [[Thursday, 11AM–9:30PM], [Friday, 11AM–9:30PM...   
2  [[Thursday, 10AM–5:30PM], [Friday, 10AM–5:30PM...   
3                                               None   
4  [[Thursday, 11AM–8PM], [Friday, 11AM–8PM], [Sa...   

                                                MISC                 state  \
0  {'Service options': ['Takeout', 'Dine-in', 'De...     Open ⋅ Closes 6PM   
1  {'Service options': ['Outdoor seating', 'Curbs...  Open ⋅ Closes 9:30PM   
2                  {'Service options': ['Delivery']}    Permanently closed   
3  {'Service options': ['Delivery'], 'Amenities':...                  None   
4  {'Service options': ['Outdoor seating', 'Takeo...   Closed ⋅ Opens 11AM   

                                    relative_results  \
0  [0x80c2c78249aba68f:0x35bf16ce61be751d, 0x80c2...   
1                                               None   
2  [0x87ee974869295555:0x95f310d065882c9b, 0x87ec...   
3  [0x89c6c62a40f4b733:0x73bfeec1898f272b, 0x89c6...   
4  [0x7c00451360f80cf1:0x930291a38bab3132, 0x7c00...   

                                                 url  
0  https://www.google.com/maps/place//data=!4m2!3...  
1  https://www.google.com/maps/place//data=!4m2!3...  
2  https://www.google.com/maps/place//data=!4m2!3...  
3  https://www.google.com/maps/place//data=!4m2!3...  
4  https://www.google.com/maps/place//data=!4m2!3...

In [13]:
#Observamos un poco el campo address
restaurants_google['address'].values

array(['San Soo Dang, 761 S Vermont Ave, Los Angeles, CA 90005',
       'Vons Chicken, 12740 La Mirada Blvd, La Mirada, CA 90638',
       'Sweet Rewards Gluten Free Bakery, LLC, 85 NE Dartmoor Dr, Waukee, IA 50263',
       ..., "Dunkin', 7000 Chester Ave, Philadelphia, PA 19142",
       'Munch, 5285 Meadows Rd #188, Lake Oswego, OR 97035',
       'Steak 48, 260 S Broad St, Philadelphia, PA 19102'], dtype=object)

Los datos están en una oración separados por una coma. Vamos a armar una lista con esto y luego tomamos los datos que nos interesan y los juntamos nuevamente.

In [26]:
#Separamos y juntamos los datos
listado_fallas = []
localizaciones = []
for i in range(restaurants_google['address'].values.size):
    try:
        int(restaurants_google['address'].values[i][-1])
        if len(restaurants_google['address'].values[i].split(',')[-1].strip()) == 8:
            localizaciones.append([restaurants_google['gmap_id'].values[i],                                     #Id del restaurant
                                restaurants_google['address'].values[i].split(',')[-2].strip(),                 #Ciudad
                                restaurants_google['address'].values[i].split(',')[-1].strip().split()[0],      #Estado
                                restaurants_google['address'].values[i].split(',')[-1].strip().split()[1]])     #Código postal
        else:
            listado_fallas.append(i)                                    #Anoto cuantas direcciones no me dan los datos necesarios
    except:
        listado_fallas.append(i)
len(listado_fallas)

12590

In [27]:
#Armamos dataframe con estos datos
localizacion = pd.DataFrame(columns=['gmap_id','Ciudad','Estado','Cod_postal'],data=localizaciones)
localizacion.head()

gmap_id        Ciudad Estado Cod_postal
0  0x80c2c778e3b73d33:0xbdc58662a4a97d49   Los Angeles     CA      90005
1  0x80dd2b4c8555edb7:0xfc33d65c4bdbef42     La Mirada     CA      90638
2   0x87ec235c54d25b31:0x3b75fb5facc602f        Waukee     IA      50263
3  0x89c6c89efcaed69d:0xded973f6033e7dba  Philadelphia     PA      19147
4   0x7c00456eecad3111:0x8217f9600c51f33          Laie     HI      96762

In [28]:
#Realizamos una comprobación. Vemos si algún valor falló
fallados = []
for i in range(localizacion.shape[0]):
    try:
        int(localizacion['Cod_postal'].iloc[i])                 #Si esto funciona quiere decir que no entraron caracteres extraños
    except:                                                     #bajo cod_postal
        print(i,': ',localizacion['Cod_postal'].iloc[i])
        fallados.append(localizacion['Cod_postal'].iloc[i])

In [29]:
#No fallo. 12590 registros no cuentan con los valores que buscamos. Los observamos a ver si se puede hacer algo.
restaurants_google[restaurants_google.index.isin(listado_fallas)]

name  \
24             Suibian Hotpot, BBQ & Bar   
189                                李小龍台吃   
378       Li's Lane 李家胡同 | 东北菜 | 烤串 | 砂锅   
396                               glazed   
414                      Japanese Cusine   
...                                  ...   
209287  Bill & Ruth's Sandwich and Sushi   
209405                     King Mackerel   
209585                        9292 Shabu   
209588                            소공동순두부   
209641               Four Sea Restaurant   

                                                  address  \
24      Connecticut, New Haven, Temple St, Suibian Hot...   
189         〒11362 New York, Queens, Northern Blvd, 李小龍台吃   
378     New York, Queens, 37th Avenue, Li's Lane 李家胡同 ...   
396                  Texas, Sugar Land, glazed邮政编码: 77479   
414                        Japanese Cusine, Georgia 30260   
...                                                   ...   
209287  74135 Oklahoma, Tulsa, S Yale Ave, Bill & Ruth...   
209405                King Mackerel, South Carolina 29812   
209585  30096 Georgia, Duluth, Old Norcross Rd, 9292 S...   
209588    36117 Alabama, Montgomery, Eastern Blvd, 소공동순두부   
209641  Pennsylvania, Harrisburg, Walnut St, Four Sea ...   

                                      gmap_id description   latitude  \
24      0x89e7d9451d24d885:0xe581040dcd508560        None  41.305142   
189     0x89c289efdb82221b:0xed627c2af97c2069        None  40.770059   
378     0x89c261968d7dab69:0xd027fc7ab6d77696        None  40.762426   
396     0x8640e6bd6e861357:0xb96bb36e5cd02ee3        None  29.590894   
414     0x88f4f95addba19ab:0x7a987482d2e71697        None  33.564651   
...                                       ...         ...        ...   
209287  0x87b6ed352c5ddd59:0x1c551b013146b7d9        None  36.117171   
209405  0x88f90effe93bcf2d:0x9f86037ad2a3a84a        None  33.272966   
209585  0x88f5a38d968d5a7f:0xaa10592ba7147e3e        None  33.961254   
209588  0x888c2a52b465553d:0xd48630befdb32560        None  32.355214   
209641  0x89c8c1bdd587ce49:0x3530fb79232f648f        None  40.275479   

        longitude                                           category  \
24     -72.927822  [Asian restaurant, Bar & grill, Hot pot restau...   
189    -73.735522                                       [Restaurant]   
378    -73.827964  [Grill, Barbecue restaurant, Chinese restauran...   
396    -95.625716                                       [Restaurant]   
414    -84.344919                                       [Restaurant]   
...           ...                                                ...   
209287 -95.922062                  [Sandwich shop, Sushi restaurant]   
209405 -81.385007                                       [Restaurant]   
209585 -84.144434                                       [Restaurant]   
209588 -86.215974                                       [Restaurant]   
209641 -76.854675  [Chinese restaurant, Asian restaurant, Mandari...   

        avg_rating  num_of_reviews price  \
24             3.7              47  None   
189            4.0              27  None   
378            4.2              18  None   
396            4.3               8  None   
414            5.0               2  None   
...            ...             ...   ...   
209287         4.4             128     $   
209405         4.5              18  None   
209585         4.3             368  None   
209588         4.4             158  None   
209641         4.8              28  None   

                                                    hours  \
24      [[Thursday, 11AM–3PM], [Friday, 11AM–3PM], [Sa...   
189     [[Wednesday, 11:30AM–10PM], [Thursday, 11:30AM...   
378     [[Wednesday, 10AM–2AM], [Thursday, 10AM–2AM], ...   
396                                                  None   
414     [[Wednesday, 11:30AM–9PM], [Thursday, 11:30AM–...   
...                                                   ...   
209287  [[Monday, 10:30AM–7PM], [Tuesday, 10:30AM–7PM]...   
209405  [[Monday, 10AM–6PM]

In [30]:
#Observamos el campo address más de cerca
list(restaurants_google[restaurants_google.index.isin(listado_fallas)]['address'].values)

['Connecticut, New Haven, Temple St, Suibian Hotpot, BBQ & Bar邮政编码: 06510',
 '〒11362 New York, Queens, Northern Blvd, 李小龍台吃',
 "New York, Queens, 37th Avenue, Li's Lane 李家胡同 | 东北菜 | 烤串 | 砂锅邮政编码: 11354",
 'Texas, Sugar Land, glazed邮政编码: 77479',
 'Japanese Cusine, Georgia 30260',
 '20194 Virginia, Reston, North Point Village Center, bb.q Chicken Reston',
 '〒95822 California, Sacramento, Freeport Blvd, 海港酒家',
 'New York, Staten Island, Victory Blvd, China sun邮政编码: 10314',
 'New York, West Islip, Higbie Ln, Panda Garden Chinese Restaurant邮政编码: 11795',
 'California, Diamond Bar, S Diamond Bar Blvd, QINJIAMO邮政编码: 91765',
 'New Jersey, Jersey City, NJ-440, Cajun Crab Shack-Jersey City邮政编码: 07304',
 '〒89103 Nevada, Las Vegas, Schiff Dr, 过桥米线',
 '90020 California, Los Angeles, W 6th St, MUN Korean Steakhouse',
 "〒96815 Hawaii, Honolulu, Kālia Rd, CJ's ニューヨーク・スタイル・デリ",
 'California, Rowland Heights, Daisetta St, Dinning Bar 酒鄉邮政编码: 91748',
 'Kirkwood Sunrise Grill, California',
 'Kentucky, Ashla

Vemos que hay muchos que tienen "United States" al final. Esto trae problemas en el algoritmo planteado. Modificando ligeramente podemos incorporar estos datos.

In [31]:
#Tomamos los valores que fallaron y los enviamos a otro dataframe
restaurants_google_fallas = restaurants_google[restaurants_google.index.isin(listado_fallas)]
restaurants_google_fallas.reset_index(inplace=True,drop=True)
restaurants_google_fallas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12590 entries, 0 to 12589
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   name              12590 non-null  object 
 1   address           12590 non-null  object 
 2   gmap_id           12590 non-null  object 
 3   description       9222 non-null   object 
 4   latitude          12590 non-null  float64
 5   longitude         12590 non-null  float64
 6   category          12590 non-null  object 
 7   avg_rating        12590 non-null  float64
 8   num_of_reviews    12590 non-null  int64  
 9   price             10799 non-null  object 
 10  hours             12225 non-null  object 
 11  MISC              12563 non-null  object 
 12  state             12249 non-null  object 
 13  relative_results  12037 non-null  object 
 14  url               12590 non-null  object 
dtypes: float64(3), int64(1), object(11)
memory usage: 1.4+ MB


In [32]:
listado_fallas = []
for i in range(restaurants_google_fallas['address'].values.size):
    try:
        lista_address = restaurants_google_fallas['address'].values[i].split(',')[0:4]      #Cortamos el "United States" de los -->
        int(lista_address[-1][-1])                                                          #registros que los tienen.
        if len(lista_address[-1].strip()) == 8:
            localizaciones.append([restaurants_google_fallas['gmap_id'].values[i],          #Id del restaurant
                                   lista_address[-2].strip(),                               #Ciudad
                                   lista_address[-1].strip().split()[0],                    #Estado
                                   lista_address[-1].strip().split()[1]])                   #Codigo postal
        else:
            listado_fallas.append(i)                                    
    except:
        listado_fallas.append(i)
len(listado_fallas)

1008

In [33]:
#Armamos dataframe con estos datos
localizacion = pd.DataFrame(columns=['gmap_id','Ciudad','Estado','Cod_postal'],data=localizaciones)
localizacion.head()

gmap_id        Ciudad Estado Cod_postal
0  0x80c2c778e3b73d33:0xbdc58662a4a97d49   Los Angeles     CA      90005
1  0x80dd2b4c8555edb7:0xfc33d65c4bdbef42     La Mirada     CA      90638
2   0x87ec235c54d25b31:0x3b75fb5facc602f        Waukee     IA      50263
3  0x89c6c89efcaed69d:0xded973f6033e7dba  Philadelphia     PA      19147
4   0x7c00456eecad3111:0x8217f9600c51f33          Laie     HI      96762

In [34]:
#Realizamos una comprobación. Vemos si algún valor falló
fallados = []
for i in range(localizacion.shape[0]):
    try:
        int(localizacion['Cod_postal'].iloc[i])                 #Si esto funciona quiere decir que no entraron caracteres extraños
    except:                                                     #bajo cod_postal
        print(i,': ',localizacion['Cod_postal'].iloc[i])
        fallados.append(localizacion['Cod_postal'].iloc[i])

In [35]:
#Se redujeron a 1008 casos nomas. Volvemos a observar.
restaurants_google_fallas[restaurants_google_fallas.index.isin(listado_fallas)]

name  \
0             Suibian Hotpot, BBQ & Bar   
1                                 李小龍台吃   
2        Li's Lane 李家胡同 | 东北菜 | 烤串 | 砂锅   
3                                glazed   
4                       Japanese Cusine   
...                                 ...   
12585  Bill & Ruth's Sandwich and Sushi   
12586                     King Mackerel   
12587                        9292 Shabu   
12588                            소공동순두부   
12589               Four Sea Restaurant   

                                                 address  \
0      Connecticut, New Haven, Temple St, Suibian Hot...   
1          〒11362 New York, Queens, Northern Blvd, 李小龍台吃   
2      New York, Queens, 37th Avenue, Li's Lane 李家胡同 ...   
3                   Texas, Sugar Land, glazed邮政编码: 77479   
4                         Japanese Cusine, Georgia 30260   
...                                                  ...   
12585  74135 Oklahoma, Tulsa, S Yale Ave, Bill & Ruth...   
12586                King Mackerel, South Carolina 29812   
12587  30096 Georgia, Duluth, Old Norcross Rd, 9292 S...   
12588    36117 Alabama, Montgomery, Eastern Blvd, 소공동순두부   
12589  Pennsylvania, Harrisburg, Walnut St, Four Sea ...   

                                     gmap_id description   latitude  \
0      0x89e7d9451d24d885:0xe581040dcd508560        None  41.305142   
1      0x89c289efdb82221b:0xed627c2af97c2069        None  40.770059   
2      0x89c261968d7dab69:0xd027fc7ab6d77696        None  40.762426   
3      0x8640e6bd6e861357:0xb96bb36e5cd02ee3        None  29.590894   
4      0x88f4f95addba19ab:0x7a987482d2e71697        None  33.564651   
...                                      ...         ...        ...   
12585  0x87b6ed352c5ddd59:0x1c551b013146b7d9        None  36.117171   
12586  0x88f90effe93bcf2d:0x9f86037ad2a3a84a        None  33.272966   
12587  0x88f5a38d968d5a7f:0xaa10592ba7147e3e        None  33.961254   
12588  0x888c2a52b465553d:0xd48630befdb32560        None  32.355214   
12589  0x89c8c1bdd587ce49:0x3530fb79232f648f        None  40.275479   

       longitude                                           category  \
0     -72.927822  [Asian restaurant, Bar & grill, Hot pot restau...   
1     -73.735522                                       [Restaurant]   
2     -73.827964  [Grill, Barbecue restaurant, Chinese restauran...   
3     -95.625716                                       [Restaurant]   
4     -84.344919                                       [Restaurant]   
...          ...                                                ...   
12585 -95.922062                  [Sandwich shop, Sushi restaurant]   
12586 -81.385007                                       [Restaurant]   
12587 -84.144434                                       [Restaurant]   
12588 -86.215974                                       [Restaurant]   
12589 -76.854675  [Chinese restaurant, Asian restaurant, Mandari...   

       avg_rating  num_of_reviews price  \
0             3.7              47  None   
1             4.0              27  None   
2             4.2              18  None   
3             4.3               8  None   
4             5.0               2  None   
...           ...             ...   ...   
12585         4.4             128     $   
12586         4.5              18  None   
12587         4.3             368  None   
12588         4.4             158  None   
12589         4.8              28  None   

                                                   hours  \
0      [[Thursday, 11AM–3PM], [Friday, 11AM–3PM], [Sa...   
1      [[Wednesday, 11:30AM–10PM], [Thursday, 11:30AM...   
2      [[Wednesday, 10AM–2AM], [Thursday, 10AM–2AM], ...   
3                                                   None   
4      [[Wednesday, 11:30AM–9PM], [Thursday, 11:30AM–...   
...                                                  ...   
12585  [[Monday, 10:30AM–7PM], [Tuesday, 10:30AM–7PM]...   
12586  [[Monday, 10AM–6PM], [Tuesday, 10AM–6PM], [Wed...   
12587  [[Monday, 11:30AM–11PM], [T

In [36]:
#Observamos el campo address mas de cerca
list(restaurants_google_fallas[restaurants_google_fallas.index.isin(listado_fallas)]['address'].values)

['Connecticut, New Haven, Temple St, Suibian Hotpot, BBQ & Bar邮政编码: 06510',
 '〒11362 New York, Queens, Northern Blvd, 李小龍台吃',
 "New York, Queens, 37th Avenue, Li's Lane 李家胡同 | 东北菜 | 烤串 | 砂锅邮政编码: 11354",
 'Texas, Sugar Land, glazed邮政编码: 77479',
 'Japanese Cusine, Georgia 30260',
 '20194 Virginia, Reston, North Point Village Center, bb.q Chicken Reston',
 '〒95822 California, Sacramento, Freeport Blvd, 海港酒家',
 'New York, Staten Island, Victory Blvd, China sun邮政编码: 10314',
 'New York, West Islip, Higbie Ln, Panda Garden Chinese Restaurant邮政编码: 11795',
 'California, Diamond Bar, S Diamond Bar Blvd, QINJIAMO邮政编码: 91765',
 'New Jersey, Jersey City, NJ-440, Cajun Crab Shack-Jersey City邮政编码: 07304',
 '〒89103 Nevada, Las Vegas, Schiff Dr, 过桥米线',
 '90020 California, Los Angeles, W 6th St, MUN Korean Steakhouse',
 "〒96815 Hawaii, Honolulu, Kālia Rd, CJ's ニューヨーク・スタイル・デリ",
 'California, Rowland Heights, Daisetta St, Dinning Bar 酒鄉邮政编码: 91748',
 'Kirkwood Sunrise Grill, California',
 'Kentucky, Ashla

Otra gran cantidad de registros tiene el código postal adelante de todo, seguido por el estado (nombre completo, no simbolo) y luego por la ciudad. Hacemos un último algoritmo para rescatar estos. Para el resto no se ve ningún patrón.

In [37]:
#Creamos un nuevo dataframe
restaurants_google_fallas2 = restaurants_google_fallas[restaurants_google_fallas.index.isin(listado_fallas)]
restaurants_google_fallas2.reset_index(inplace=True,drop=True)
restaurants_google_fallas2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1008 entries, 0 to 1007
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   name              1008 non-null   object 
 1   address           1008 non-null   object 
 2   gmap_id           1008 non-null   object 
 3   description       275 non-null    object 
 4   latitude          1008 non-null   float64
 5   longitude         1008 non-null   float64
 6   category          1008 non-null   object 
 7   avg_rating        1008 non-null   float64
 8   num_of_reviews    1008 non-null   int64  
 9   price             399 non-null    object 
 10  hours             917 non-null    object 
 11  MISC              1001 non-null   object 
 12  state             918 non-null    object 
 13  relative_results  724 non-null    object 
 14  url               1008 non-null   object 
dtypes: float64(3), int64(1), object(11)
memory usage: 118.2+ KB


In [38]:
#Primero cambiamos los nombres de los estados por sus códigos
state_codes = {'Alabama': 'AL','Alaska': 'AK','Arizona': 'AZ','Arkansas': 'AR','California': 'CA','Colorado': 'CO','Connecticut': 'CT',
    'Delaware': 'DE','Florida': 'FL','Georgia': 'GA','Hawaii': 'HI','Idaho': 'ID','Illinois': 'IL','Indiana': 'IN','Iowa': 'IA',
    'Kansas': 'KS','Kentucky': 'KY','Louisiana': 'LA','Maine': 'ME','Maryland': 'MD','Massachusetts': 'MA','Michigan': 'MI',
    'Minnesota': 'MN','Mississippi': 'MS','Missouri': 'MO','Montana': 'MT','Nebraska': 'NE','Nevada': 'NV','New Hampshire': 'NH',
    'New Jersey': 'NJ','New Mexico': 'NM','New York': 'NY','North Carolina': 'NC','North Dakota': 'ND','Ohio': 'OH','Oklahoma': 'OK',
    'Oregon': 'OR','Pennsylvania': 'PA','Rhode Island': 'RI','South Carolina': 'SC','South Dakota': 'SD','Tennessee': 'TN','Texas': 'TX',
    'Utah': 'UT','Vermont': 'VT','Virginia': 'VA','Washington': 'WA','West Virginia': 'WV','Wisconsin': 'WI','Wyoming': 'WY'}

for state, code in state_codes.items():
    restaurants_google_fallas2['address'] = restaurants_google_fallas2['address'].str.replace(state,code)
restaurants_google_fallas2

C:\Users\larra\AppData\Local\Temp\ipykernel_15676\4027301428.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  restaurants_google_fallas2['address'] = restaurants_google_fallas2['address'].str.replace(state,code)


name  \
0            Suibian Hotpot, BBQ & Bar   
1                                李小龍台吃   
2       Li's Lane 李家胡同 | 东北菜 | 烤串 | 砂锅   
3                               glazed   
4                      Japanese Cusine   
...                                ...   
1003  Bill & Ruth's Sandwich and Sushi   
1004                     King Mackerel   
1005                        9292 Shabu   
1006                            소공동순두부   
1007               Four Sea Restaurant   

                                                address  \
0     CT, New Haven, Temple St, Suibian Hotpot, BBQ ...   
1               〒11362 NY, Queens, Northern Blvd, 李小龍台吃   
2     NY, Queens, 37th Avenue, Li's Lane 李家胡同 | 东北菜 ...   
3                     TX, Sugar Land, glazed邮政编码: 77479   
4                             Japanese Cusine, GA 30260   
...                                                 ...   
1003  74135 OK, Tulsa, S Yale Ave, Bill & Ruth's San...   
1004                            King Mackerel, SC 29812   
1005      30096 GA, Duluth, Old Norcross Rd, 9292 Shabu   
1006         36117 AL, Montgomery, Eastern Blvd, 소공동순두부   
1007  PA, Harrisburg, Walnut St, Four Sea Restaurant...   

                                    gmap_id description   latitude  longitude  \
0     0x89e7d9451d24d885:0xe581040dcd508560        None  41.305142 -72.927822   
1     0x89c289efdb82221b:0xed627c2af97c2069        None  40.770059 -73.735522   
2     0x89c261968d7dab69:0xd027fc7ab6d77696        None  40.762426 -73.827964   
3     0x8640e6bd6e861357:0xb96bb36e5cd02ee3        None  29.590894 -95.625716   
4     0x88f4f95addba19ab:0x7a987482d2e71697        None  33.564651 -84.344919   
...                                     ...         ...        ...        ...   
1003  0x87b6ed352c5ddd59:0x1c551b013146b7d9        None  36.117171 -95.922062   
1004  0x88f90effe93bcf2d:0x9f86037ad2a3a84a        None  33.272966 -81.385007   
1005  0x88f5a38d968d5a7f:0xaa10592ba7147e3e        None  33.961254 -84.144434   
1006  0x888c2a52b465553d:0xd48630befdb32560        None  32.355214 -86.215974   
1007  0x89c8c1bdd587ce49:0x3530fb79232f648f        None  40.275479 -76.854675   

                                               category  avg_rating  \
0     [Asian restaurant, Bar & grill, Hot pot restau...         3.7   
1                                          [Restaurant]         4.0   
2     [Grill, Barbecue restaurant, Chinese restauran...         4.2   
3                                          [Restaurant]         4.3   
4                                          [Restaurant]         5.0   
...                                                 ...         ...   
1003                  [Sandwich shop, Sushi restaurant]         4.4   
1004                                       [Restaurant]         4.5   
1005                                       [Restaurant]         4.3   
1006                                       [Restaurant]         4.4   
1007  [Chinese restaurant, Asian restaurant, Mandari...         4.8   

      num_of_reviews price                                              hours  \
0                 47  None  [[Thursday, 11AM–3PM], [Friday, 11AM–3PM], [Sa...   
1                 27  None  [[Wednesday, 11:30AM–10PM], [Thursday, 11:30AM...   
2                 18  None  [[Wednesday, 10AM–2AM], [Thursday, 10AM–2AM], ...   
3                  8  None                                               None   
4                  2  None  [[Wednesday, 11:30AM–9PM], [Thursday, 11:30AM–...   
...              ...   ...                                                ...   
1003             128     $  [[Monday, 10:30AM–7PM], [Tuesday, 10:30AM–7PM]...   
1004              18  None  [[Monday, 10AM–6PM], [Tuesday, 10AM–6PM], [Wed...   
1005             368  None  [[Monday, 11:30AM–11PM], [Tuesday, 11:30AM–11P...   
1006             158  None  [[Monday, 10:30AM–10PM], [Tuesday, 10:30AM–10P...   
1007              28  None  [[Monday, Closed], [Tuesday, 11AM–8PM], [Wedne...   

                     

In [39]:
#Quitamos los caracteres extraños
for i in range(restaurants_google_fallas2.shape[0]):
    restaurants_google_fallas2['address'][i] = re.sub(r'[^a-zA-Z0-9,\s]', '', restaurants_google_fallas2['address'][i])
restaurants_google_fallas2.head()

C:\Users\larra\AppData\Local\Temp\ipykernel_15676\4089036508.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  restaurants_google_fallas2['address'][i] = re.sub(r'[^a-zA-Z0-9,\s]', '', restaurants_google_fallas2['address'][i])


name  \
0       Suibian Hotpot, BBQ & Bar   
1                           李小龍台吃   
2  Li's Lane 李家胡同 | 东北菜 | 烤串 | 砂锅   
3                          glazed   
4                 Japanese Cusine   

                                             address  \
0  CT, New Haven, Temple St, Suibian Hotpot, BBQ ...   
1                  11362 NY, Queens, Northern Blvd,    
2     NY, Queens, 37th Avenue, Lis Lane        11354   
3                       TX, Sugar Land, glazed 77479   
4                          Japanese Cusine, GA 30260   

                                 gmap_id description   latitude  longitude  \
0  0x89e7d9451d24d885:0xe581040dcd508560        None  41.305142 -72.927822   
1  0x89c289efdb82221b:0xed627c2af97c2069        None  40.770059 -73.735522   
2  0x89c261968d7dab69:0xd027fc7ab6d77696        None  40.762426 -73.827964   
3  0x8640e6bd6e861357:0xb96bb36e5cd02ee3        None  29.590894 -95.625716   
4  0x88f4f95addba19ab:0x7a987482d2e71697        None  33.564651 -84.344919   

                                            category  avg_rating  \
0  [Asian restaurant, Bar & grill, Hot pot restau...         3.7   
1                                       [Restaurant]         4.0   
2  [Grill, Barbecue restaurant, Chinese restauran...         4.2   
3                                       [Restaurant]         4.3   
4                                       [Restaurant]         5.0   

   num_of_reviews price                                              hours  \
0              47  None  [[Thursday, 11AM–3PM], [Friday, 11AM–3PM], [Sa...   
1              27  None  [[Wednesday, 11:30AM–10PM], [Thursday, 11:30AM...   
2              18  None  [[Wednesday, 10AM–2AM], [Thursday, 10AM–2AM], ...   
3               8  None                                               None   
4               2  None  [[Wednesday, 11:30AM–9PM], [Thursday, 11:30AM–...   

                                                MISC  \
0  {'Service options': ['Delivery'], 'Accessibili...   
1  {'Service options': ['Takeout', 'Dine-in', 'De...   
2  {'From the business': ['Identifies as women-le...   
3  {'Service options': ['Delivery'], 'Amenities':...   
4  {'Service options': ['Takeout', 'Dine-in', 'De...   

                        state  \
0          Permanently closed   
1  Closed ⋅ Opens 11:30AM Thu   
2           Open ⋅ Closes 2AM   
3                        None   
4           Open ⋅ Closes 9PM   

                                    relative_results  \
0  [0x89e7d84b4c3ac797:0x9bf376c7e3132709, 0x89e7...   
1  [0x89c289efe0d5d521:0xb30eff4172713bd6, 0x89c2...   
2                                               None   
3                                               None   
4  [0x88f4f930e95db43f:0x94f550e5a38f7948, 0x88f4...   

                                                 url  
0  https://www.google.com/maps/place//data=!4m2!3...  
1  https://www.google.com/maps/place//data=!4m2!3...  
2  https://www.google.com/maps/place//data=!4m2!3...  
3  https://www.google.com/maps/place//data=!4m2!3...  
4  https://www.google.com/maps/place//data=!4m2!3...

In [40]:
listado_fallas = []
for i in range(restaurants_google_fallas2['address'].values.size):
    try:
        int(restaurants_google_fallas2.iloc[i]['address'][0])         #Si esto sucede entonces el primer digito es un numero
        localizaciones.append([restaurants_google_fallas2['gmap_id'].values[i],                                     #Id del restaurant
                            restaurants_google_fallas2['address'].values[i].split(',')[1].strip(),                 #Ciudad
                            restaurants_google_fallas2['address'].values[i].split(',')[0].strip().split()[1],      #Estado
                            restaurants_google_fallas2['address'].values[i].split(',')[0].strip().split()[0]])     #Código postal
    except:
        listado_fallas.append(i)                                        #Anoto cuantas direcciones no me dan los datos necesarios
len(listado_fallas)

777

In [41]:
#Armamos dataframe con estos datos
localizacion = pd.DataFrame(columns=['gmap_id','Ciudad','Estado','Cod_postal'],data=localizaciones)
localizacion.head()

gmap_id        Ciudad Estado Cod_postal
0  0x80c2c778e3b73d33:0xbdc58662a4a97d49   Los Angeles     CA      90005
1  0x80dd2b4c8555edb7:0xfc33d65c4bdbef42     La Mirada     CA      90638
2   0x87ec235c54d25b31:0x3b75fb5facc602f        Waukee     IA      50263
3  0x89c6c89efcaed69d:0xded973f6033e7dba  Philadelphia     PA      19147
4   0x7c00456eecad3111:0x8217f9600c51f33          Laie     HI      96762

In [42]:
#Realizamos una comprobación. Vemos si algún valor falló
fallados = []
for i in range(localizacion.shape[0]):
    try:
        int(localizacion['Cod_postal'].iloc[i])                 #Si esto funciona quiere decir que no entraron caracteres extraños
    except:                                                     #bajo cod_postal
        print(i,': ',localizacion['Cod_postal'].iloc[i])
        fallados.append(i)

Redujimos a 777 los casos de los que no podemos obtener los datos necesarios. Vamos a juntar este dataframe con el original.

In [43]:
#Juntamos con el de google
restaurants_google = pd.merge(google,localizacion,how='left',on='gmap_id')
restaurants_google.head()

name  \
0                           San Soo Dang   
1                           Vons Chicken   
2  Sweet Rewards Gluten Free Bakery, LLC   
3                        Vivi Bubble Tea   
4                              Hale Pops   

                                             address  \
0  San Soo Dang, 761 S Vermont Ave, Los Angeles, ...   
1  Vons Chicken, 12740 La Mirada Blvd, La Mirada,...   
2  Sweet Rewards Gluten Free Bakery, LLC, 85 NE D...   
3  Vivi Bubble Tea, 701 S 5th St, Philadelphia, P...   
4   Hale Pops, 55-370 Kamehameha Hwy, Laie, HI 96762   

                                 gmap_id  \
0  0x80c2c778e3b73d33:0xbdc58662a4a97d49   
1  0x80dd2b4c8555edb7:0xfc33d65c4bdbef42   
2   0x87ec235c54d25b31:0x3b75fb5facc602f   
3  0x89c6c89efcaed69d:0xded973f6033e7dba   
4   0x7c00456eecad3111:0x8217f9600c51f33   

                                         description   latitude   longitude  \
0                                               None  34.058092 -118.292130   
1                                               None  33.916402 -118.010855   
2                                               None  41.616079  -93.865487   
3  Bubble tea shop offering hot & cold versions, ...  39.940293  -75.150923   
4                                               None  21.637796 -157.920714   

                           category  avg_rating  num_of_reviews price  \
0               [Korean restaurant]         4.4              18  None   
1                      [Restaurant]         4.5              18  None   
2  [Bakery, Health food restaurant]         4.7              21  None   
3                      [Restaurant]         4.0               8  None   
4                      [Restaurant]         4.4              18  None   

                                               hours  \
0  [[Thursday, 6:30AM–6PM], [Friday, 6:30AM–6PM],...   
1  [[Thursday, 11AM–9:30PM], [Friday, 11AM–9:30PM...   
2  [[Thursday, 10AM–5:30PM], [Friday, 10AM–5:30PM...   
3                                               None   
4  [[Thursday, 11AM–8PM], [Friday, 11AM–8PM], [Sa...   

                                                MISC                 state  \
0  {'Service options': ['Takeout', 'Dine-in', 'De...     Open ⋅ Closes 6PM   
1  {'Service options': ['Outdoor seating', 'Curbs...  Open ⋅ Closes 9:30PM   
2                  {'Service options': ['Delivery']}    Permanently closed   
3  {'Service options': ['Delivery'], 'Amenities':...                  None   
4  {'Service options': ['Outdoor seating', 'Takeo...   Closed ⋅ Opens 11AM   

                                    relative_results  \
0  [0x80c2c78249aba68f:0x35bf16ce61be751d, 0x80c2...   
1                                               None   
2  [0x87ee974869295555:0x95f310d065882c9b, 0x87ec...   
3  [0x89c6c62a40f4b733:0x73bfeec1898f272b, 0x89c6...   
4  [0x7c00451360f80cf1:0x930291a38bab3132, 0x7c00...   

                                                 url        Ciudad Estado  \
0  https://www.google.com/maps/place//data=!4m2!3...   Los Angeles     CA   
1  https://www.google.com/maps/place//data=!4m2!3...     La Mirada     CA   
2  https://www.google.com/maps/place//data=!4m2!3...        Waukee     IA   
3  https://www.google.com/maps/place//data=!4m2!3...  Philadelphia     PA   
4  https://www.google.com/maps/place//data=!4m2!3...          Laie     HI   

  Cod_postal  
0      90005  
1      90638  
2      50263  
3      19147  
4      96762

In [44]:
restaurants_google.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 210953 entries, 0 to 210952
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   name              210953 non-null  object 
 1   address           209900 non-null  object 
 2   gmap_id           210953 non-null  object 
 3   description       72174 non-null   object 
 4   latitude          210953 non-null  float64
 5   longitude         210953 non-null  float64
 6   category          210953 non-null  object 
 7   avg_rating        210953 non-null  float64
 8   num_of_reviews    210953 non-null  int64  
 9   price             99738 non-null   object 
 10  hours             189013 non-null  object 
 11  MISC              209549 non-null  object 
 12  state             189848 non-null  object 
 13  relative_results  170277 non-null  object 
 14  url               210953 non-null  object 
 15  Ciudad            209123 non-null  object 
 16  Estado            20

Terminada esta parte vamos a armar el dataframe que deseamos. Descartamos las columnas que no sirven, cambiamos nombre, reordenamos, etc.

In [45]:
#Descartamos columnas
restaurants_google.drop(columns=['description','avg_rating','num_of_reviews','price',
                                 'hours','state','relative_results','url','address'], inplace=True)
restaurants_google.head()

name  \
0                           San Soo Dang   
1                           Vons Chicken   
2  Sweet Rewards Gluten Free Bakery, LLC   
3                        Vivi Bubble Tea   
4                              Hale Pops   

                                 gmap_id   latitude   longitude  \
0  0x80c2c778e3b73d33:0xbdc58662a4a97d49  34.058092 -118.292130   
1  0x80dd2b4c8555edb7:0xfc33d65c4bdbef42  33.916402 -118.010855   
2   0x87ec235c54d25b31:0x3b75fb5facc602f  41.616079  -93.865487   
3  0x89c6c89efcaed69d:0xded973f6033e7dba  39.940293  -75.150923   
4   0x7c00456eecad3111:0x8217f9600c51f33  21.637796 -157.920714   

                           category  \
0               [Korean restaurant]   
1                      [Restaurant]   
2  [Bakery, Health food restaurant]   
3                      [Restaurant]   
4                      [Restaurant]   

                                                MISC        Ciudad Estado  \
0  {'Service options': ['Takeout', 'Dine-in', 'De...   Los Angeles     CA   
1  {'Service options': ['Outdoor seating', 'Curbs...     La Mirada     CA   
2                  {'Service options': ['Delivery']}        Waukee     IA   
3  {'Service options': ['Delivery'], 'Amenities':...  Philadelphia     PA   
4  {'Service options': ['Outdoor seating', 'Takeo...          Laie     HI   

  Cod_postal  
0      90005  
1      90638  
2      50263  
3      19147  
4      96762

In [46]:
#Cambiamos nombre
restaurants_google.rename(columns={'name':'Nombre','gmap_id':'Id_Restaurant','latitude':'Latitud','longitude':'Longitud',
                                   'category':'Tipo','MISC':'Atributos'},inplace=True)
restaurants_google.head()

Nombre  \
0                           San Soo Dang   
1                           Vons Chicken   
2  Sweet Rewards Gluten Free Bakery, LLC   
3                        Vivi Bubble Tea   
4                              Hale Pops   

                           Id_Restaurant    Latitud    Longitud  \
0  0x80c2c778e3b73d33:0xbdc58662a4a97d49  34.058092 -118.292130   
1  0x80dd2b4c8555edb7:0xfc33d65c4bdbef42  33.916402 -118.010855   
2   0x87ec235c54d25b31:0x3b75fb5facc602f  41.616079  -93.865487   
3  0x89c6c89efcaed69d:0xded973f6033e7dba  39.940293  -75.150923   
4   0x7c00456eecad3111:0x8217f9600c51f33  21.637796 -157.920714   

                               Tipo  \
0               [Korean restaurant]   
1                      [Restaurant]   
2  [Bakery, Health food restaurant]   
3                      [Restaurant]   
4                      [Restaurant]   

                                           Atributos        Ciudad Estado  \
0  {'Service options': ['Takeout', 'Dine-in', 'De...   Los Angeles     CA   
1  {'Service options': ['Outdoor seating', 'Curbs...     La Mirada     CA   
2                  {'Service options': ['Delivery']}        Waukee     IA   
3  {'Service options': ['Delivery'], 'Amenities':...  Philadelphia     PA   
4  {'Service options': ['Outdoor seating', 'Takeo...          Laie     HI   

  Cod_postal  
0      90005  
1      90638  
2      50263  
3      19147  
4      96762

In [47]:
#Reordenamos
restaurants_google = restaurants_google[['Id_Restaurant','Nombre','Ciudad','Estado','Cod_postal', 
                                         'Latitud', 'Longitud', 'Tipo', 'Atributos']]
restaurants_google.head()

Id_Restaurant  \
0  0x80c2c778e3b73d33:0xbdc58662a4a97d49   
1  0x80dd2b4c8555edb7:0xfc33d65c4bdbef42   
2   0x87ec235c54d25b31:0x3b75fb5facc602f   
3  0x89c6c89efcaed69d:0xded973f6033e7dba   
4   0x7c00456eecad3111:0x8217f9600c51f33   

                                  Nombre        Ciudad Estado Cod_postal  \
0                           San Soo Dang   Los Angeles     CA      90005   
1                           Vons Chicken     La Mirada     CA      90638   
2  Sweet Rewards Gluten Free Bakery, LLC        Waukee     IA      50263   
3                        Vivi Bubble Tea  Philadelphia     PA      19147   
4                              Hale Pops          Laie     HI      96762   

     Latitud    Longitud                              Tipo  \
0  34.058092 -118.292130               [Korean restaurant]   
1  33.916402 -118.010855                      [Restaurant]   
2  41.616079  -93.865487  [Bakery, Health food restaurant]   
3  39.940293  -75.150923                      [Restaurant]   
4  21.637796 -157.920714                      [Restaurant]   

                                           Atributos  
0  {'Service options': ['Takeout', 'Dine-in', 'De...  
1  {'Service options': ['Outdoor seating', 'Curbs...  
2                  {'Service options': ['Delivery']}  
3  {'Service options': ['Delivery'], 'Amenities':...  
4  {'Service options': ['Outdoor seating', 'Takeo...

In [48]:
#Observamos el resultado
restaurants_google.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 210953 entries, 0 to 210952
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Id_Restaurant  210953 non-null  object 
 1   Nombre         210953 non-null  object 
 2   Ciudad         209123 non-null  object 
 3   Estado         209123 non-null  object 
 4   Cod_postal     209123 non-null  object 
 5   Latitud        210953 non-null  float64
 6   Longitud       210953 non-null  float64
 7   Tipo           210953 non-null  object 
 8   Atributos      209549 non-null  object 
dtypes: float64(2), object(7)
memory usage: 16.1+ MB


Hay 1830 valores nulos de los campos Ciudad, Estado y Cod_Postal. Es menos del 1% del dataset y el trabajo para completar esto es demasiado. Vamos a descartar estos registros.

In [49]:
#Descartamos
restaurants_google.dropna(subset='Cod_postal',inplace=True)
restaurants_google.reset_index(inplace=True,drop=True)
restaurants_google.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209123 entries, 0 to 209122
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Id_Restaurant  209123 non-null  object 
 1   Nombre         209123 non-null  object 
 2   Ciudad         209123 non-null  object 
 3   Estado         209123 non-null  object 
 4   Cod_postal     209123 non-null  object 
 5   Latitud        209123 non-null  float64
 6   Longitud       209123 non-null  float64
 7   Tipo           209123 non-null  object 
 8   Atributos      207894 non-null  object 
dtypes: float64(2), object(7)
memory usage: 14.4+ MB


In [50]:
#Vamos a observar si hay negocios que se encuentren duplicados.
#Es decir que tengan distinto Id (ya se descartó antes los de igual Id) pero coincidan nombre y ubicación.
duplicados_google = restaurants_google[restaurants_google.duplicated(subset=['Nombre','Cod_postal'],keep=False)].sort_values(by='Nombre')
duplicados_google.head()

Id_Restaurant          Nombre         Ciudad  \
10900  0x89c2f5d10275956b:0x678f175621275540  1 Dollar Pizza          Bronx   
8038   0x89c2f556a7d50035:0x1588f291f277f706  1 Dollar Pizza          Bronx   
67508  0x88d9b76485f1a105:0x92bc2888314d41f4  2 Korean Girls  Coconut Grove   
5705   0x88d9b1e28cc6f4db:0x4dc90cfb7b4f01b7  2 Korean Girls  Coconut Grove   
43806  0x88ded74449b67443:0x6f5ce881bf9883d0        3Natives        Jupiter   

      Estado Cod_postal    Latitud   Longitud  \
10900     NY      10451  40.818870 -73.927612   
8038      NY      10451  40.818870 -73.927612   
67508     FL      33133  25.729263 -80.240075   
5705      FL      33133  25.729263 -80.240075   
43806     FL      33458  26.935733 -80.135056   

                                                    Tipo  \
10900                                 [Pizza restaurant]   
8038                                        [Restaurant]   
67508                                [Korean restaurant]   
5705                                 [Korean restaurant]   
43806  [Health food restaurant, Cafe, Juice shop, San...   

                                               Atributos  
10900  {'Service options': ['Delivery'], 'Popular for...  
8038   {'Service options': ['Takeout', 'Delivery'], '...  
67508  {'Service options': ['Curbside pickup', 'No-co...  
5705   {'From the business': ['Identifies as women-le...  
43806  {'Service options': ['Delivery', 'Takeout', 'D...

In [51]:
#Vemos que hay y tienen la misma ubicación. Vemos cuantos son.
duplicados_google.reset_index(inplace=True,drop=True)
duplicados_google.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7936 entries, 0 to 7935
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id_Restaurant  7936 non-null   object 
 1   Nombre         7936 non-null   object 
 2   Ciudad         7936 non-null   object 
 3   Estado         7936 non-null   object 
 4   Cod_postal     7936 non-null   object 
 5   Latitud        7936 non-null   float64
 6   Longitud       7936 non-null   float64
 7   Tipo           7936 non-null   object 
 8   Atributos      7925 non-null   object 
dtypes: float64(2), object(7)
memory usage: 558.1+ KB


In [52]:
#Vemos cuantos negocios son
negocios_duplicados_google = duplicados_google.drop_duplicates(subset=['Nombre','Cod_postal'])
negocios_duplicados_google.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3515 entries, 0 to 7934
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id_Restaurant  3515 non-null   object 
 1   Nombre         3515 non-null   object 
 2   Ciudad         3515 non-null   object 
 3   Estado         3515 non-null   object 
 4   Cod_postal     3515 non-null   object 
 5   Latitud        3515 non-null   float64
 6   Longitud       3515 non-null   float64
 7   Tipo           3515 non-null   object 
 8   Atributos      3512 non-null   object 
dtypes: float64(2), object(7)
memory usage: 274.6+ KB


Son 3515 negocios que forman 7936 registros. Vamos a quedarnos con uno de los duplicados y descartar los otros. Sin embargo, vamos a guardar los ids así no perdemos información de las reseñas. Luego usaremos esta info para modificar los indices de las reseñas de google.

In [53]:
#Tomamos los ids a conservar y los que se van a cambiar
lista_duplicados = []
for n in negocios_duplicados_google['Id_Restaurant'].values:
    aux = []
    nombre = negocios_duplicados_google[negocios_duplicados_google['Id_Restaurant']==n]['Nombre'].values[0]
    cod_postal = negocios_duplicados_google[negocios_duplicados_google['Id_Restaurant']==n]['Cod_postal'].values[0]
    for d in range(duplicados_google.shape[0]):
        if (duplicados_google['Nombre'][d] == nombre) and (duplicados_google['Cod_postal'][d] == cod_postal) and (duplicados_google['Id_Restaurant'][d] != n):
            aux.append(duplicados_google['Id_Restaurant'][d])
    lista_duplicados.append([n,aux])

In [54]:
#Armammos dataframe
duplicados_google_df = pd.DataFrame(data=lista_duplicados,columns=['Conservar','Cambiar'])
duplicados_google_df.head()

Conservar  \
0  0x89c2f5d10275956b:0x678f175621275540   
1  0x88d9b76485f1a105:0x92bc2888314d41f4   
2  0x88ded74449b67443:0x6f5ce881bf9883d0   
3  0x89b60170d06531f1:0xcb85dc5e00da9e1f   
4  0x89c259a11e60603d:0x92f338b57ed650bb   

                                   Cambiar  
0  [0x89c2f556a7d50035:0x1588f291f277f706]  
1  [0x88d9b1e28cc6f4db:0x4dc90cfb7b4f01b7]  
2  [0x88ded58b9368604d:0x4fb1d504b6ae7eee]  
3  [0x89b60170d0e68baf:0x67a51b3fda1b9d5e]  
4  [0x89c2597fb6fffff9:0xd6e4a37d0c56f328]

In [55]:
#Descartamos los duplicados
restaurants_google.drop_duplicates(subset=['Nombre','Cod_postal'],inplace=True)
restaurants_google.reset_index(inplace=True,drop=True)
restaurants_google.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 204702 entries, 0 to 204701
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Id_Restaurant  204702 non-null  object 
 1   Nombre         204702 non-null  object 
 2   Ciudad         204702 non-null  object 
 3   Estado         204702 non-null  object 
 4   Cod_postal     204702 non-null  object 
 5   Latitud        204702 non-null  float64
 6   Longitud       204702 non-null  float64
 7   Tipo           204702 non-null  object 
 8   Atributos      203478 non-null  object 
dtypes: float64(2), object(7)
memory usage: 14.1+ MB


Redujimos de más de 3.000.000 de registros a, solamente, alrededor de 200.000. Los datos descartados no contenían información relevante. Vamos a exportar 3 dataframes importantes:

* restaurants_google: Es el dataset con el de restaurants de google con el que vamos a trabajar.
* duplicados_google_df: Es el dataset de valores duplicados que hay que cambiar en la review.
* google: Es el dataset auxiliar con el que vamos a hacer un primer filtro en las review para hacer el cambio con duplicados_google_df en un datasest mas liviano.

In [56]:
#Exportamos a formato pickle y csv
restaurants_google.to_pickle('metadata-sitios/restaurants_google.pkl')
google.to_pickle('metadata-sitios/google.pkl')
duplicados_google_df.to_csv('metadata-sitios/duplicados.csv')

Ahora continuaremos con la normalización de los valores del dataframe. Los campos con valores no normalizados que quedan son Tipo y Atributos.

In [3]:
#Importamos el dataset final de los restaurants de google
restaurants_google = pd.read_pickle('metadata-sitios/restaurants_google.pkl')
restaurants_google.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123911 entries, 0 to 123910
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Id_Restaurant  123911 non-null  object 
 1   Nombre         123911 non-null  object 
 2   Ciudad         123911 non-null  object 
 3   Estado         123911 non-null  object 
 4   Cod_postal     123911 non-null  object 
 5   Latitud        123911 non-null  float64
 6   Longitud       123911 non-null  float64
 7   Tipo           123911 non-null  object 
 8   Atributos      123450 non-null  object 
dtypes: float64(2), object(7)
memory usage: 8.5+ MB


In [11]:
tipos = []
for t in list(restaurants_google['Tipo'].values):
    tipos += t
tipos = set(tipos)
len(tipos)

1391

In [12]:
tipos

{'Hawaiian restaurant',
 'Wildlife and safari park',
 'Tofu restaurant',
 'Artificial plant supplier',
 'Southwestern restaurant (US)',
 'State liquor store',
 'Hookah bar',
 'Kerosene supplier',
 'Outdoor furniture store',
 'Food producer',
 'Magician',
 'Gay bar',
 'Brake shop',
 'Heating oil supplier',
 'Fax service',
 'Food and beverage consultant',
 'Champon noodle restaurant',
 'Outlet store',
 'Pizza takeaway',
 'South American restaurant',
 'Performing arts theater',
 'Travel',
 'Bakery',
 'Cruise agency',
 'Junk store',
 'Haunted house',
 'Kaiseki restaurant',
 'Madrilian restaurant',
 'Furniture accessories',
 'Pizza Takeout',
 'Culinary school',
 'Group accommodation',
 'Stand bar',
 'Recording studio',
 'Asian household goods store',
 'Home improvement store',
 'Buffet restaurant',
 'Indonesian restaurant',
 'Fruit wholesaler',
 'Vineyard',
 'Diesel engine repair service',
 'New Zealand restaurant',
 'Confectionery wholesaler',
 'Furnished apartment building',
 'Jamaican re